### ISSUE TITLE
Threading/Isolate issue ->Spontaneous JVM crash on creation of V8

### ISSUE THREAD

**Issue by matiwinnetou** [eclipsesource/J2V8#306]
Let me explain a bit setup.

We have a new page running, 100% J2V8 on the server with react and react on the client.

This renderer consists of J2V8 with apache commons pool and evictor running every 10 minutes. There is minimum 5 j2v8 that thread has to keep in pool. Generally it works but every now and then JVM keeps crashing.

```
drwxr-xr-x  3 root   root             4096 Nov 10  2015 ..
mszczap@pubse47-15:/var/tmp/backend-server/pubse47-15_b01_3606$ ls -lat *.log
-rw-rw-rw- 1 mobile mobile_local 212586 Jul 13 11:55 hs_err_pid9112.log
-rw-rw-rw- 1 mobile mobile_local  95270 Jul 11 08:52 hs_err_pid9816.log
-rw-rw-rw- 1 mobile mobile_local  89220 Jul 11 05:31 hs_err_pid2823.log
-rw-rw-rw- 1 mobile mobile_local 211672 Jul 10 06:20 hs_err_pid21594.log
-rw-rw-rw- 1 mobile mobile_local 109212 Jul  7 19:09 hs_err_pid9514.log
-rw-rw-rw- 1 mobile mobile_local 220100 Jul  4 22:36 hs_err_pid10330.log
-rw-rw-rw- 1 mobile mobile_local  98499 Jul  4 20:05 hs_err_pid8388.log
-rw-rw-rw- 1 mobile mobile_local 106891 Jul  4 19:04 hs_err_pid2222.log
```

and here info on stack trace, seems to be in native code

```
---------------  T H R E A D  ---------------

Current thread (0x00007f3224780800):  JavaThread "commons-pool-EvictionTimer" daemon [_thread_in_native, id=10358, stack(0x00007f3271027000,0x00007f3271128000)]

siginfo: si_signo: 11 (SIGSEGV), si_code: 1 (SEGV_MAPERR), si_addr: 0x0000000000000000

Register to memory mapping:

RAX=0x0000000000000000 is an unknown value
RBX=0x00007f321c261710 is an unknown value
RCX=0x00007f3299f0c2e0: <offset 0x2182e0> in /lib/x86_64-linux-gnu/libpthread.so.0 at 0x00007f3299cf4000
RDX=0x00007f3271127a50 is pointing into the stack for thread: 0x00007f3224780800
RSP=0x00007f32711260c0 is pointing into the stack for thread: 0x00007f3224780800
RBP=0x00007f32711260e0 is pointing into the stack for thread: 0x00007f3224780800
RSI=0x0000000000000001 is an unknown value
RDI=0x0000000000000040 is an unknown value
R8 =0x0000000000000002 is an unknown value
R9 =0x00007f321c082ff0 is an unknown value
R10=0x0000000000000001 is an unknown value
R11=0x00000000173aaf7d is an unknown value
R12=0x00007f321c24d6d0 is an unknown value
R13=0x00007f321c261710 is an unknown value
R14=0x00007f32711264b0 is pointing into the stack for thread: 0x00007f3224780800
R15=0x00007f3224780800 is a thread


Stack: [0x00007f3271027000,0x00007f3271128000],  sp=0x00007f32711260c0,  free space=1020k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [libj2v8_linux_x86_64.so+0x8ca0c7]  v8::internal::Isolate::Enter()+0x87
C  [libj2v8_linux_x86_64.so+0x546281]  v8::Isolate::Scope::Scope(v8::Isolate*)+0x27
C  [libj2v8_linux_x86_64.so+0x52c92b]  Java_com_eclipsesource_v8_V8__1createIsolate+0xde
j  com.eclipsesource.v8.V8._createIsolate(Ljava/lang/String;)J+0
j  com.eclipsesource.v8.V8.<init>(Ljava/lang/String;)V+87
j  com.eclipsesource.v8.V8.createV8Runtime(Ljava/lang/String;Ljava/lang/String;)Lcom/eclipsesource/v8/V8;+56
j  pubse.ecs.system.renderer.servlet.j2v8.V8Factory.createV8()Lcom/eclipsesource/v8/V8;+7
j  pubse.ecs.system.renderer.servlet.j2v8.V8PooledObjectPageFactory.makeObject()Lorg/apache/commons/pool2/PooledObject;+0
j  org.apache.commons.pool2.impl.GenericObjectPool.create()Lorg/apache/commons/pool2/PooledObject;+47
j  org.apache.commons.pool2.impl.GenericObjectPool.ensureIdle(IZ)V+39
j  org.apache.commons.pool2.impl.GenericObjectPool.ensureMinIdle()V+6
j  org.apache.commons.pool2.impl.BaseGenericObjectPool$Evictor.run()V+89
j  java.util.TimerThread.mainLoop()V+221
j  java.util.TimerThread.run()V+1
v  ~StubRoutines::call_stub
V  [libjvm.so+0x681a56]
V  [libjvm.so+0x681f61]
V  [libjvm.so+0x682407]
V  [libjvm.so+0x7182b0]
V  [libjvm.so+0xa5c96f]
V  [libjvm.so+0xa5ca9c]
V  [libjvm.so+0x910ee8]
C  [libpthread.so.0+0x80a4]  start_thread+0xc4
```

Make object is acquiring lock and releasing after v8 is created, all boxes have 5 GB ram for which 1-2 GB is reserved for JVM and the rest for native memory (V8). Since it was risky for us to move to J2V8 we decided to host a new use case on new set of boxes in production and observe what happens. On various machines we see sporadic crashes from time to time.

Commons pool evictor for the moment is ran every 10 mins and destroys and then creates J2V8 based on traffic on demand.

Make object is implemented as follows:
```
    @Override
    public PooledObject<CompiledV8Bundle> makeObject() throws Exception {
        V8 v8 = createV8();
        V8Locker locker = v8.getLocker();
        try {
            locker.acquire();
            logger.info("makeV8, thread:" + locker.getThread().getName());
            CompiledV8Bundle compiledV8Bundle = compileV8Bundle(v8, compileTimeGlobals, assetMapping);

            return new DefaultPooledObject<>(compiledV8Bundle);
        } finally {
            locker.release();
        }
    }
```

We observe those crashes on at least two JVM 1.8 versions:
1.8.0_40 and 1.8.0_131 64 bit on intel both running on linux ubuntu (xenial)

Any idea what could cause this? Is this some bug in our code or more likely in native C++ JNI bridge to V8?

Last but not least, the version of j2v8 we use is 4.8.0.

**Comment by matiwinnetou** [comment]
We observe those crashes on at least two JVM 1.8 versions:
java -version
java version "1.8.0_40"
Java(TM) SE Runtime Environment (build 1.8.0_40-b25)
Java HotSpot(TM) 64-Bit Server VM (build 25.40-b25, mixed mode)

and also on version: _131 (linux)

**Comment by matiwinnetou** [comment]
Analysing this further I can reach a conclusion that there is some probability higher than 0 for JVM crash (perhaps race condition) upon creation of V8.

**Comment by matiwinnetou** [comment]
@irbull As a workaround is it possible to catch this native exception in the java land and then perhaps try to create V8 again? Crash of JVM sounds serious.

**Comment by matiwinnetou** [comment]
We see more exceptions, logs to follow. What are the recommendations in terms of JVM settings to run J2V8? -XSS param, etc?

**Comment by matiwinnetou** [comment]
There are more sporadic errors request bound:
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f13561f2a8d, pid=39, tid=0x00007f134cbf7700
#
# JRE version: Java(TM) SE Runtime Environment (8.0_131-b11) (build 1.8.0_131-b11)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.131-b11 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
#
# Core dump written. Default location: //core or core.39
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

---------------  T H R E A D  ---------------

Current thread (0x00007f1230002000):  JavaThread "http-nio-8080-exec-9" daemon [_thread_in_native, id=146, stack(0x00007f134caf7000,0x00007f134cbf8000)]

siginfo: si_signo: 11 (SIGSEGV), si_code: 128 (SI_KERNEL), si_addr: 0x0000000000000000

Registers:
RAX=0x303035313a227461, RBX=0x00007f134cbf5700, RCX=0x00007f13e62a02c0, RDX=0x00007f134cbf7a30
RSP=0x00007f134cbf56a0, RBP=0x00007f134cbf56c0, RSI=0x0000000000000001, RDI=0x0000000000000020
R8 =0x0000000000000003, R9 =0x00001f703a104241, R10=0x00001f703a1aadb9, R11=0x00007f13e5a38f50
R12=0x0000000000000000, R13=0x00000000e50cc6c0, R14=0x00007f1268078ef0, R15=0x00007f1230002000
RIP=0x00007f13561f2a8d, EFLAGS=0x0000000000010246, CSGSFS=0x0000000000000033, ERR=0x0000000000000000
  TRAPNO=0x000000000000000d

Top of Stack: (sp=0x00007f134cbf56a0)
0x00007f134cbf56a0:   00007f134cbf56c0 00007f13562146d1
0x00007f134cbf56b0:   00007f126807aab0 00007f126807aab0
0x00007f134cbf56c0:   00007f134cbf58d0 00007f135613aa04
0x00007f134cbf56d0:   00007f134cbf5a30 00007f134cbf58f0
0x00007f134cbf56e0:   00007f134cbf58f8 00007f134cbf5908
0x00007f134cbf56f0:   00007f12ec008390 00007f12300021f8
0x00007f134cbf5700:   303035313a227461 0000000000000000
0x00007f134cbf5710:   0000000400000003 0000000068060f00
0x00007f134cbf5720:   00007f1300000060 00007f12ec008390
0x00007f134cbf5730:   00007f1268060ff8 0000000000000000
0x00007f134cbf5740:   00007f1268060fe8 00007f1268061000
0x00007f134cbf5750:   00007f1268061038 ffffffff00000002
0x00007f134cbf5760:   00007f134cbf5790 00007f1268060fe8
0x00007f134cbf5770:   00007f1268061038 00007f1268060ff0
0x00007f134cbf5780:   00007f1268061038 00007f1230002000
0x00007f134cbf5790:   00007f1268060ff0 00007f135621471d
0x00007f134cbf57a0:   00007f1268060fe8 00007f135614d29f
0x00007f134cbf57b0:   00007f1268060fe8 00007f134cbf58b0
0x00007f134cbf57c0:   00007f134cbf5a20 00007f1356145f28
0x00007f134cbf57d0:   00007f1268060fe8 00007f12ec05c290
0x00007f134cbf57e0:   00007f1268060ff0 00007f12ec008330
0x00007f134cbf57f0:   00007f134cbf5a30 00007f12300021f8
0x00007f134cbf5800:   00007f1268060ff8 000000000000000b
0x00007f134cbf5810:   00007f1268061038 00007f1268061038
0x00007f134cbf5820:   00007f1268061038 00007f126807aab0
0x00007f134cbf5830:   00007f1268060ff8 00007f1268060ff8
0x00007f134cbf5840:   00007f12ec062770 00007f12ec062770
0x00007f134cbf5850:   00007f12ec062770 00007f12ec008390
0x00007f134cbf5860:   00007f1268078ef0 00007f12ec008390
0x00007f134cbf5870:   00007f1268060ff0 00007f12ec0626d0
0x00007f134cbf5880:   00007f12ec0626d0 00007f12ec0626d0
0x00007f134cbf5890:   00007f12ec008390 00007f12ec05bad0 

Instructions: (pc=0x00007f13561f2a8d)
0x00007f13561f2a6d:   f3 ff 90 55 48 89 e5 53 48 89 fb 48 83 ec 18 48
0x00007f13561f2a7d:   8b 05 a5 2c 03 01 8b 38 e8 e6 b5 f0 ff 48 89 03
0x00007f13561f2a8d:   48 8b 80 e0 8c 00 00 48 89 df 48 89 43 08 0f b6
0x00007f13561f2a9d:   43 28 83 e0 e0 83 c8 06 88 43 28 e8 f3 6c f3 ff 

Register to memory mapping:

RAX=0x303035313a227461 is an unknown value
RBX=0x00007f134cbf5700 is pointing into the stack for thread: 0x00007f1230002000
RCX=0x00007f13e62a02c0: <offset 0x2182c0> in /lib/x86_64-linux-gnu/libpthread.so.0 at 0x00007f13e6088000
RDX=0x00007f134cbf7a30 is pointing into the stack for thread: 0x00007f1230002000
RSP=0x00007f134cbf56a0 is pointing into the stack for thread: 0x00007f1230002000
RBP=0x00007f134cbf56c0 is pointing into the stack for thread: 0x00007f1230002000
RSI=0x0000000000000001 is an unknown value
RDI=0x0000000000000020 is an unknown value
R8 =0x0000000000000003 is an unknown value
R9 =0x00001f703a104241 is an unknown value
R10=0x00001f703a1aadb9 is an unknown value
R11=0x00007f13e5a38f50: <offset 0x194f50> in /lib/x86_64-linux-gnu/libc.so.6 at 0x00007f13e58a4000
R12=0x0000000000000000 is an unknown value
R13=0x00000000e50cc6c0 is an oop
[Ljava.lang.Object; 
 - klass: 'java/lang/Object'[]
 - length: 3
R14=0x00007f1268078ef0 is an unknown value
R15=0x00007f1230002000 is a thread


Stack: [0x00007f134caf7000,0x00007f134cbf8000],  sp=0x00007f134cbf56a0,  free space=1017k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
C  [libj2v8_linux_x86_64.so+0x533a04]  invokeFunction(JNIEnv_*, v8::Isolate*, long&, long&, _jstring*&, long&, v8::Local<v8::Value>&)+0x300
C  [libj2v8_linux_x86_64.so+0x533ed4]  Java_com_eclipsesource_v8_V8__1executeFunction__JIJLjava_lang_String_2J+0x192
J 10028  com.eclipsesource.v8.V8._executeFunction(JIJLjava/lang/String;J)Ljava/lang/Object; (0 bytes) @ 0x00007f13d2134e15 [0x00007f13d2134d40+0xd5]
J 11377 C2 com.eclipsesource.v8.V8Object.executeJSFunction(Ljava/lang/String;[Ljava/lang/Object;)Ljava/lang/Object; (279 bytes) @ 0x00007f13d193f188 [0x00007f13d193ee40+0x348]
J 11579 C2 pubse.ecs.system.renderer.servlet.j2v8.PoolingJ2V8JavaScriptRenderer.apply(Lde/mobile/ecs/World;)V (340 bytes) @ 0x00007f13d242bc8c [0x00007f13d242b420+0x86c]
J 11594 C2 pubse.ecs.system.renderer.ResponseRenderer.apply(Lde/mobile/ecs/World;)V (79 bytes) @ 0x00007f13d2441740 [0x00007f13d24413a0+0x3a0]
J 11147 C2 com.google.inject.internal.InterceptorStackCallback$InterceptedMethodInvocation.proceed()Ljava/lang/Object; (91 bytes) @ 0x00007f13d23a8b64 [0x00007f13d23a2c60+0x5f04]
J 11512 C2 pubse.config.InboundModule$Monitor.invoke(Lorg/aopalliance/intercept/MethodInvocation;)Ljava/lang/Object; (70 bytes) @ 0x00007f13d240ae1c [0x00007f13d240ad00+0x11c]
J 11147 C2 com.google.inject.internal.InterceptorStackCallback$InterceptedMethodInvocation.proceed()Ljava/lang/Object; (91 bytes) @ 0x00007f13d23a2de0 [0x00007f13d23a2c60+0x180]
J 11738 C2 pubse.tomcat.handler.ServletRouter$$Lambda$185.handle(Ljavax/servlet/http/HttpServletRequest;Ljavax/servlet/http/HttpServletResponse;)V (10 bytes) @ 0x00007f13d254dd38 [0x00007f13d254d500+0x838]
J 12185 C2 org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;)V (574 bytes) @ 0x00007f13d11a9104 [0x00007f13d11a8fa0+0x164]
J 11186 C2 org.ocpsoft.rewrite.servlet.RewriteFilter.doFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;Ljavax/servlet/FilterChain;)V (402 bytes) @ 0x00007f13d23c854c [0x00007f13d23c7200+0x134c]
J 12185 C2 org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;)V (574 bytes) @ 0x00007f13d11a940c [0x00007f13d11a8fa0+0x46c]
J 11550 C2 pubse.util.ServletFilters$Utf8EncodingFilter.doFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;Ljavax/servlet/FilterChain;)V (21 bytes) @ 0x00007f13d2399980 [0x00007f13d23995a0+0x3e0]
J 12185 C2 org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;)V (574 bytes) @ 0x00007f13d11a940c [0x00007f13d11a8fa0+0x46c]
J 11623 C2 pubse.util.ServletFilters$LoggingFilter.doFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;Ljavax/servlet/FilterChain;)V (76 bytes) @ 0x00007f13d245dd6c [0x00007f13d245da40+0x32c]
J 12185 C2 org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;)V (574 bytes) @ 0x00007f13d11a940c [0x00007f13d11a8fa0+0x46c]
J 11543 C2 org.apache.catalina.core.StandardWrapperValve.invoke(Lorg/apache/catalina/connector/Request;Lorg/apache/catalina/connector/Response;)V (1389 bytes) @ 0x00007f13d1dc6f2c [0x00007f13d1dc6d40+0x1ec]
J 11538 C2 org.apache.catalina.core.StandardContextValve.invoke(Lorg/apache/catalina/connector/Request;Lorg/apache/catalina/connector/Response;)V (172 bytes) @ 0x00007f13d1e035ec [0x00007f13d1e03060+0x58c]
J 11851 C2 org.apache.catalina.authenticator.AuthenticatorBase.invoke(Lorg/apache/catalina/connector/Request;Lorg/apache/catalina/connector/Response;)V (903 bytes) @ 0x00007f13d207a6ec [0x00007f13d207a2c0+0x42c]
J 11733 C2 org.apache.catalina.core.StandardHostValve.invoke(Lorg/apache/catalina/connector/Request;Lorg/apache/catalina/connector/Response;)V (402 bytes) @ 0x00007f13d25469b0 [0x00007f13d2546640+0x370]
J 11584 C2 org.apache.catalina.connector.CoyoteAdapter.service(Lorg/apache/coyote/Request;Lorg/apache/coyote/Response;)V (805 bytes) @ 0x00007f13d2445554 [0x00007f13d2445060+0x4f4]
J 11808 C2 org.apache.coyote.http11.AbstractHttp11Processor.process(Lorg/apache/tomcat/util/net/SocketWrapper;)Lorg/apache/tomcat/util/net/AbstractEndpoint$Handler$SocketState; (1128 bytes) @ 0x00007f13d25793f8 [0x00007f13d2579180+0x278]
J 11805 C2 org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(Lorg/apache/tomcat/util/net/SocketWrapper;Lorg/apache/tomcat/util/net/SocketStatus;)Lorg/apache/tomcat/util/net/AbstractEndpoint$Handler$SocketState; (1073 bytes) @ 0x00007f13d1f4a6dc [0x00007f13d1f4a200+0x4dc]
J 11444 C2 org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun()V (603 bytes) @ 0x00007f13d18b1250 [0x00007f13d18b10e0+0x170]
J 11438 C2 org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run()V (79 bytes) @ 0x00007f13d19f216c [0x00007f13d19f20c0+0xac]
j  java.util.concurrent.ThreadPoolExecutor.runWorker(Ljava/util/concurrent/ThreadPoolExecutor$Worker;)V+95
j  java.util.concurrent.ThreadPoolExecutor$Worker.run()V+5
j  org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run()V+4
j  java.lang.Thread.run()V+11
v  ~StubRoutines::call_stub
V  [libjvm.so+0x691d16]  JavaCalls::call_helper(JavaValue*, methodHandle*, JavaCallArguments*, Thread*)+0x1056
V  [libjvm.so+0x692221]  JavaCalls::call_virtual(JavaValue*, KlassHandle, Symbol*, Symbol*, JavaCallArguments*, Thread*)+0x321
V  [libjvm.so+0x6926c7]  JavaCalls::call_virtual(JavaValue*, Handle, KlassHandle, Symbol*, Symbol*, Thread*)+0x47
V  [libjvm.so+0x72da50]  thread_entry(JavaThread*, Thread*)+0xa0
V  [libjvm.so+0xa76833]  JavaThread::thread_main_inner()+0x103
V  [libjvm.so+0xa7697c]  JavaThread::run()+0x11c
V  [libjvm.so+0x927568]  java_start(Thread*)+0x108
C  [libpthread.so.0+0x76ba]  start_thread+0xca
```

**Comment by matiwinnetou** [comment]
```mszczap@shell47-1:~$ grep -ir -A1 "Problematic frame:" .
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x60c617]  v8::Isolate::GetCurrentContext()+0x27
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x60c5f0]  v8::Isolate::GetCurrentContext()+0x0
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x60c5f0]  v8::Isolate::GetCurrentContext()+0x0
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x60c5f0]  v8::Isolate::GetCurrentContext()+0x0
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x60c5f0]  v8::Isolate::GetCurrentContext()+0x0
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x60c5f0]  v8::Isolate::GetCurrentContext()+0x0
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x60c5f0]  v8::Isolate::GetCurrentContext()+0x0
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x60c5f0]  v8::Isolate::GetCurrentContext()+0x0
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x60c5f0]  v8::Isolate::GetCurrentContext()+0x0
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x60c5f0]  v8::Isolate::GetCurrentContext()+0x0
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x60c5fc]  v8::Isolate::GetCurrentContext()+0xc
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x5eba8d]  v8::TryCatch::TryCatch()+0x1d
--
./exceptions.tmp:# Problematic frame:
./exceptions.tmp-# C  [libj2v8_linux_x86_64.so+0x60c5f0]  v8::Isolate::GetCurrentContext()+0x0

**Comment by matiwinnetou** [comment]
So in total there are three errors that we found that result in jvm crash:
1. sometimes on creation of v8, reported above
2. v8::TryCatch::TryCatch()
3. v8::Isolate::GetCurrentContext()

**Comment by matiwinnetou** [comment]
[j2v8_exceptions.txt](https://github.com/eclipsesource/J2V8/files/1152042/j2v8_exceptions.txt)

**Comment by matiwinnetou** [comment]
@irbull it looks like they are fatal exceptions in V8? Any reasons why they happen? Do you know if there is a way to catch them to prevent JVM exiting? Any other hints that can help us?

**Comment by matiwinnetou** [comment]
Since this issue is in fact I think possibly three different bugs I will most likely split this into 3 issues.

We can relatively reliably reproduce JVM crash on creation of V8 engine (happens in various test systems we have - sporadically) and happens after few hundred times of instance creation. I am now trying to reproduce other crashes as unit tests.

**Comment by matiwinnetou** [comment]
To some extend those issues do not surprise me, J2V8 has so far been primarily used for UI development and we operate a cluster of servers. Some errors occur only under heavy load with lots of users hitting (we have 1.500 req/s to the cluster). No bug can easily slip through this.

**Comment by matiwinnetou** [comment]
I am not a C++ developer but looking at the code:
```

**Comment by matiwinnetou** [comment]
Another observation is that there seem to be two TryCatch constructors one with *Isolate and another one without. We are using the one without *Isolate

```
v8::TryCatch::TryCatch()
    : isolate_(i::Isolate::Current()),
      next_(isolate_->try_catch_handler()),
      is_verbose_(false),
      can_continue_(true),
      capture_message_(true),
      rethrow_(false),
      has_terminated_(false) {
  ResetInternal();
  // Special handling for simulators which have a separate JS stack.
  js_stack_comparable_address_ =
      reinterpret_cast<void*>(v8::internal::SimulatorStack::RegisterCTryCatch(
          isolate_, v8::internal::GetCurrentStackPosition()));
  isolate_->RegisterTryCatchHandler(this);
}


v8::TryCatch::TryCatch(v8::Isolate* isolate)
    : isolate_(reinterpret_cast<i::Isolate*>(isolate)),
      next_(isolate_->try_catch_handler()),
      is_verbose_(false),
      can_continue_(true),
      capture_message_(true),
      rethrow_(false),
      has_terminated_(false) {
  ResetInternal();
  // Special handling for simulators which have a separate JS stack.
  js_stack_comparable_address_ =
      reinterpret_cast<void*>(v8::internal::SimulatorStack::RegisterCTryCatch(
          isolate_, v8::internal::GetCurrentStackPosition()));
  isolate_->RegisterTryCatchHandler(this);
}
```

**Comment by matiwinnetou** [comment]
We sat down couple of guys and hypothesis is that this is issue with lack of passing Isolate, so we are changing to:

TryCatch tryCatch(isolate);

and we will build and try in production

**Comment by matiwinnetou** [comment]
another hypothesis is that there is not enough stack size for exception but error log says that there is almost 1MB stack size left.

**Comment by matiwinnetou** [comment]
Looks like Isolate::Current or Isolate::GetCurrent is deprecated and anti-pattern:

Isolate* Isolate::New(const Isolate::CreateParams& params) {
  i::Isolate* isolate = new i::Isolate(false);
  Isolate* v8_isolate = reinterpret_cast<Isolate*>(isolate);
  CHECK(params.array_buffer_allocator != NULL);
  isolate->set_array_buffer_allocator(params.array_buffer_allocator);
  if (params.snapshot_blob != NULL) {
    isolate->set_snapshot_blob(params.snapshot_blob);
  } else {
    isolate->set_snapshot_blob(i::Snapshot::DefaultSnapshotBlob());
  }
  if (params.entry_hook) {
    isolate->set_function_entry_hook(params.entry_hook);
  }
  auto code_event_handler = params.code_event_handler;
#ifdef ENABLE_GDB_JIT_INTERFACE
  if (code_event_handler == nullptr && i::FLAG_gdbjit) {
    code_event_handler = i::GDBJITInterface::EventHandler;
  }
#endif  // ENABLE_GDB_JIT_INTERFACE
  if (code_event_handler) {
    isolate->InitializeLoggingAndCounters();
    isolate->logger()->SetCodeEventHandler(kJitCodeEventDefault,
                                           code_event_handler);
  }
  if (params.counter_lookup_callback) {
    v8_isolate->SetCounterFunction(params.counter_lookup_callback);
  }

  if (params.create_histogram_callback) {
    v8_isolate->SetCreateHistogramFunction(params.create_histogram_callback);
  }

  if (params.add_histogram_sample_callback) {
    v8_isolate->SetAddHistogramSampleFunction(
        params.add_histogram_sample_callback);
  }

  isolate->set_api_external_references(params.external_references);
  SetResourceConstraints(isolate, params.constraints);
  // TODO(jochen): Once we got rid of Isolate::Current(), we can remove this.
  Isolate::Scope isolate_scope(v8_isolate);
  if (params.entry_hook || !i::Snapshot::Initialize(isolate)) {
    isolate->Init(NULL);
  }
  return v8_isolate;
}

**Comment by matiwinnetou** [comment]
Most likely this exception is happening somewhere here:
```
void Isolate::Enter() {
  Isolate* current_isolate = NULL;
  PerIsolateThreadData* current_data = CurrentPerIsolateThreadData();
  if (current_data != NULL) {
    current_isolate = current_data->isolate_;
    DCHECK(current_isolate != NULL);
    if (current_isolate == this) {
      DCHECK(Current() == this);
      DCHECK(entry_stack_ != NULL);
      DCHECK(entry_stack_->previous_thread_data == NULL ||
             entry_stack_->previous_thread_data->thread_id().Equals(
                 ThreadId::Current()));
      // Same thread re-enters the isolate, no need to re-init anything.
      entry_stack_->entry_count++;
      return;
    }
  }

  PerIsolateThreadData* data = FindOrAllocatePerThreadDataForThisThread();
  DCHECK(data != NULL);
  DCHECK(data->isolate_ == this);

  EntryStackItem* item = new EntryStackItem(current_data,
                                            current_isolate,
                                            entry_stack_);
  entry_stack_ = item;

  SetIsolateThreadLocals(this, data);

  // In case it's the first time some thread enters the isolate.
  set_thread_id(data->thread_id());
}
```

**Comment by matiwinnetou** [comment]
The way we are able to reproduce is the following, we create and destroy J2V8 in a pool by 20 threads, previously allocating locks accordingly.

```
15:09:36,623 [commons-pool-EvictionTimer] INFO  p.e.s.r.s.j.V8PooledObjectPageFactory   makeV8, thread:commons-pool-EvictionTimer
15:09:46,135 [commons-pool-EvictionTimer] INFO  p.e.s.r.s.j.V8PooledObjectPageFactory   destroyV8, thread:commons-pool-EvictionTimer
15:09:46,138 [commons-pool-EvictionTimer] INFO  p.e.s.r.s.j.V8PooledObjectPageFactory   destroyV8, thread:commons-pool-EvictionTimer
15:09:46,140 [commons-pool-EvictionTimer] INFO  p.e.s.r.s.j.V8PooledObjectPageFactory   destroyV8, thread:commons-pool-EvictionTimer
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00000001254a8624, pid=47554, tid=0x0000000000005f03
#
# JRE version: Java(TM) SE Runtime Environment (8.0_112-b16) (build 1.8.0_112-b16)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.112-b16 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [libj2v8_macosx_x86_64.dylib+0x517624]  v8::internal::Isolate::Enter()+0x34
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/mszczap/Devel/mobile/public-search-germany-webapp/hs_err_pid47554.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

Process finished with exit code 134 (interrupted by signal 6: SIGABRT)
```

**Comment by matiwinnetou** [comment]
I got a core file, it has 26 GB though:
W git:(upstream ⚡ j2v8_crash_init) 16≡ 1M ls -lah /cores/core.81741
-r--------  1 mszczap  admin    26G Jul 19 20:07 /cores/core.81741

**Comment by matiwinnetou** [comment]
```
void Isolate::Enter() {
  printf("1\n");
  fflush(stdout);
  Isolate* current_isolate = NULL;
  PerIsolateThreadData* current_data = CurrentPerIsolateThreadData();
  if (current_data != NULL) {
    printf("2\n");
  fflush(stdout);
    current_isolate = current_data->isolate_;
    DCHECK(current_isolate != NULL);
    printf("3\n");
  fflush(stdout);
    if (current_isolate == this) {
      printf("4\n");
  fflush(stdout);
      DCHECK(Current() == this);
      DCHECK(entry_stack_ != NULL);
      printf("5\n");
  fflush(stdout);
      DCHECK(entry_stack_->previous_thread_data == NULL ||
             entry_stack_->previous_thread_data->thread_id().Equals(
                 ThreadId::Current()));

      printf("6\n");
  fflush(stdout);
      // Same thread re-enters the isolate, no need to re-init anything.
      entry_stack_->entry_count++;

      return;
    }
  }

  printf("7\n");
  fflush(stdout);

  PerIsolateThreadData* data = FindOrAllocatePerThreadDataForThisThread();

  printf("8\n");
  fflush(stdout);

  DCHECK(data != NULL);
  DCHECK(data->isolate_ == this);

  printf("9\n");
  fflush(stdout);

  EntryStackItem* item = new EntryStackItem(current_data,
                                            current_isolate,
                                            entry_stack_);
  entry_stack_ = item;

  printf("10\n");
  fflush(stdout);

  SetIsolateThreadLocals(this, data);

  printf("11\n");
  fflush(stdout);

  // In case it's the first time some thread enters the isolate.
  set_thread_id(data->thread_id());

  printf("12\n");
  fflush(stdout);
}
```

**Comment by matiwinnetou** [comment]
```
20:51:39,671 [commons-pool-EvictionTimer] INFO  p.e.s.r.s.j.V8PooledObjectPageFactory   destroyV8, thread:commons-pool-EvictionTimer
20:51:39,676 [commons-pool-EvictionTimer] INFO  p.e.s.r.s.j.V8PooledObjectPageFactory   destroyV8, thread:commons-pool-EvictionTimer
20:51:39,679 [commons-pool-EvictionTimer] INFO  p.e.s.r.s.j.V8PooledObjectPageFactory   destroyV8, thread:commons-pool-EvictionTimer
1
2
3
4
5
6
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x0000000126d1a5ac, pid=1969, tid=0x0000000000006103
#
# JRE version: Java(TM) SE Runtime Environment (8.0_131-b11) (build 1.8.0_131-b11)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.131-b11 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [libj2v8_macosx_x86_64.dylib+0x5175ac]  v8::internal::Isolate::Enter()+0xbc
#
# Core dump written. Default location: /cores/core or core.1969
#
# An error report file with more information is saved as:
# /Users/mszczap/Devel/mobile/public-search-germany-webapp/hs_err_pid1969.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```

**Comment by matiwinnetou** [comment]
So clearly 6a is missing..., which means that line that causes crash is:
```
      // Same thread re-enters the isolate, no need to re-init anything.
      entry_stack_->entry_count++;
```

Even the latest V8 has this code in master:
https://github.com/v8/v8/blob/2580f56342c658284be928cb65813cbaeb53aff2/src/isolate.cc

**Comment by matiwinnetou** [comment]
I will try to run the build with assertions on.

**Comment by matiwinnetou** [comment]
With assertions on it is not even possible to debug...

```

#
# Fatal error in ../deps/v8/src/api.h, line 490
# Check failed: blocks_.length() == 0.
#

==== C stack trace ===============================

#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGILL (0x4) at pc=0x000000012b862001, pid=84230, tid=0x0000000000005f03
#
# JRE version: Java(TM) SE Runtime Environment (8.0_112-b16) (build 1.8.0_112-b16)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.112-b16 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [libj2v8_macosx_x86_64.dylib+0x1018001]  _ZN2v84base2OS5AbortEv+0x11
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/mszczap/Devel/mobile/public-search-germany-webapp/hs_err_pid84230.log
    0   libj2v8_macosx_x86_64.dylib         0x000000012b85d84e v8::base::debug::StackTrace::StackTrace() + 30
    1   libj2v8_macosx_x86_64.dylib         0x000000012b85d885 v8::base::debug::StackTrace::StackTrace() + 21
    2   libj2v8_macosx_x86_64.dylib         0x000000012b856124 V8_Fatal + 452
    3   libj2v8_macosx_x86_64.dylib         0x000000012a9aaaa2 v8::internal::HandleScopeImplementer::Free() + 98
    4   libj2v8_macosx_x86_64.dylib         0x000000012a9aaa35 v8::internal::HandleScopeImplementer::FreeThreadResources() + 21
    5   libj2v8_macosx_x86_64.dylib         0x000000012b68c774 v8::internal::ThreadManager::FreeThreadResources() + 276
    6   libj2v8_macosx_x86_64.dylib         0x000000012b68c622 v8::Locker::~Locker() + 146
    7   libj2v8_macosx_x86_64.dylib         0x000000012b68ca95 v8::Locker::~Locker() + 21
    8   libj2v8_macosx_x86_64.dylib         0x000000012a939215 Java_com_eclipsesource_v8_V8__1createIsolate + 3317
    9   ???                                 0x000000010c93e9f4 0x0 + 4505987572
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

Process finished with exit code 134 (interrupted by signal 6: SIGABRT)
```

**Comment by irbull** [comment]
I just returned from my summer vacation. I'm going through my backlog of emails, bugs, slack messages, etc.. I see you raised / commented on a bunch of issues @matiwinnetou. I will begin looking through these today.

**Comment by matiwinnetou** [comment]
@irbull Thank you so much, I am willing to support you and help. I have been digging deep to this but my C++ knowledge is really rusty I must say. I provided lots of exceptions and stack traces and I hope insights into the problem

**Comment by matiwinnetou** [comment]
I have been thinking that perhaps while in prod this bug does not produce many issues with easy workarounds of not recreating v8s very often I have to say that this bug maybe in fact hides a real problem that we are facing in #313.

Anyway, for your info, I am able to reproduce this bug on a dev box (OSX, Sierra) on latest branch : "isolate_enter_branch" build quite quickly (as before).

**Comment by matiwinnetou** [comment]
I will try to test this as well, tomorrow or on Monday if latest master build fixes this...

**Comment by matiwinnetou** [comment]
I tested this now number of minutes, I cannot reproduce this anymore. This seems fixed. I will reopen when I discover this again.
